### PR TITLE
Introduce support for API idempotency

### PR DESCRIPTION
**Pull request by piotrchowaniec**

### CODE REVIEW

File: src/main/java/com/shift4/Shift4Gateway.java [hunk 1]
```diff
@@ -62,10 +62,18 @@ public Charge createCharge(ChargeRequest request) {
         return post(CHARGES_PATH, request, Charge.class);
     }
 
+    public Charge createCharge(ChargeRequest request, RequestOptions requestOptions) {
+        return post(CHARGES_PATH, request, requestOptions, Charge.class);
+    }
```
Comment (el-dot) [inline src/main/java/com/shift4/Shift4Gateway.java:67]: We should add idempotency key to ChargeRequest as json ignored and use it in client to add header to not multiply methods

**Review by el-dot — CHANGES_REQUESTED** [review]

File: src/main/java/com/shift4/Shift4Gateway.java (hunk 1 above)
Comment (piotrchowaniec) [reply src/main/java/com/shift4/Shift4Gateway.java:67]: RequestOptions is quite a standard way of doing this kind of feature. I wouldn't say I like the idea, of messing up the request body with some additional data.

Comment (el-dot) [reply src/main/java/com/shift4/Shift4Gateway.java:67]: Ok, lets leave this as separate param then

File: src/main/java/com/shift4/request/RequestOptions.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    }
+
+    public RequestOptions() {
+    }
+
+    public String getIdempotencyKey() {
+        return idempotencyKey;
+    }
+
+    public RequestOptions setIdempotencyKey(String idempotencyKey) {
```
[the remark this answers is not in the document]
Comment (piotrchowaniec) [reply src/main/java/com/shift4/request/RequestOptions.java:25]: Hmm, this is a fluent setter.
Also getter is used in Shift4GatewayHeadersFactory.

File: src/main/java/com/shift4/request/RequestOptions.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+public class RequestOptions {
+    private String idempotencyKey;
+
+    public static RequestOptions withIdempotencyKey(String idempotencyKey) {
+        RequestOptions requestOptions = new RequestOptions();
+        requestOptions.idempotencyKey = idempotencyKey;
+        return requestOptions;
+    }
+
+    public static RequestOptions withIdempotencyKey() {
```
[the remark this answers is not in the document]
Comment (piotrchowaniec) [reply src/main/java/com/shift4/request/RequestOptions.java]: Ok, will remove

File: src/main/java/com/shift4/request/RequestOptions.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (el-dot) [reply src/main/java/com/shift4/request/RequestOptions.java:19]: Depends on definition. Following lombok nomenclature:
`X#setIdempotencyKey(String): X` is chainable and fluent is the name without set `X#idempotencyKey(String)`

We use both for setters: setter without set in name and chainable.

[the remark this answers is not in the document]
Comment (el-dot) [reply src/main/java/com/shift4/request/RequestOptions.java:19]: Getter is used in factory indeed, we can leave it in a such case.

[the remark this answers is not in the document]
Comment (piotrchowaniec) [reply src/main/java/com/shift4/request/RequestOptions.java:19]: done

**Pull request merged**